Review

SmolinIvan opened the pull request.

SmolinIvan committed:
Slices for order (feed) and ingredients. Added render for /feed and main page

SmolinIvan committed:
feat: slices, auth and protected route
Какой-то огромный коммит
1. Добавил слайс для пользователя (регистрация, авторизация, работа с токенами)
2. Частично настроил редиректы

SmolinIvan committed:
feat: Added modals with current routing
1. Добавил открытие и закрытие модалок для фида и ингридиентов
2. При переходе по прямому урлу ингридиента открывается страница с открытой модалкой нужного ингридиента
3. Закрытие модалок реализовано через две соответствующие функции (навигация на нуржный роут)

SmolinIvan committed:
feat: slices fix, order information and modal added
1. Порефачил слайсы
2. Добавил поле для ошибки при авторизации
3. Донастроил редиректы для неавторизованного пользователя
4. Доделал флоу оформления заказа: Клик на оформить -> Открытие модалки со статусом (либо редирект) -> Модалка с информацией заказа
5. Сделал отоброжение модалки с заказом с определенного заказа

SmolinIvan committed:
feat: fixed imports for slices path, added error for Register
1. Настроил экспорты для слайсов
2. Добавил ошибку при регистрации на страницу регистрации

SmolinIvan committed:
feat: Functionality checked, added path for store, added lodout extraReducer
1. Перенс logout из экшенов в экстраредьюсеры
2. Настроил авторизацию сразу после регистрации
3. Добавил путь для стора
4. Проверил весь функционал

mikebars1995 commented on src/services/slices/orderSlice.ts:
```diff
@@ -0,0 +1,178 @@
+      prepare: (ingredient: TIngredient) => ({
+        payload: {
+          ...ingredient,
+          id: nanoid()
```
**Отлично**

* Отлично, что в `prepare` создаете `id`

mikebars1995 commented on src/services/slices/userSlice.ts:
```diff
@@ -0,0 +1,182 @@
+        state.accessToken = action.payload.accessToken;
+        state.refreshToken = action.payload.refreshToken;
+        state.errorText = '';
+        setCookie('accessToken', state.accessToken);
```
**Нужно исправить**

Редьюсер должен быть **чистой функцией**. Устанавливать куки в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией.

mikebars1995 commented on src/components/burger-ingredients/burger-ingredients.tsx:
```diff
@@ -1,14 +1,16 @@
-import { TTabMode } from '@utils-types';
+import { TIngredient, TTabMode } from '@utils-types';
 import { BurgerIngredientsUI } from '../ui/burger-ingredients';
+import { useDispatch, useSelector } from 'react-redux';
```
**Нужно исправить**

* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что в `prepare` создаете `id`
* Хорошая логика попапов
* Отлично сделана логика отображения попапа и страницы  под ним
* Хорошо сделали `ProtectedRoute`


но есть некоторые недочеты:


* Редьюсер должен быть **чистой функцией**. Устанавливать куки в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. 
* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.


**Можно лучше**

* при клике на **логотип** обычно идет переход на **главную** страницу

SmolinIvan committed:
fix: fixed reducers, added link to BurgerIcon, fixed imports
1. Поправил импорты useDispatch и useSelector
2. Поправил логику редьюсеров, вынес установку и удаление токенов в AsyncThunk
3. Добавил редирект на иконку бургера

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

SmolinIvan committed:
feat: added Cypress for project
1. Установил и настроил сайпрес
2. Добавил стартовый тест

SmolinIvan committed:
feat: added Cypress tests

SmolinIvan committed:
feat: Added tests for actions and async actions. Jest configured
1. Настроил импорты для jest
2. Добавил несоклько тестов
3. Переключаюсб на настройку запуска тестов в GitHub Actions

SmolinIvan committed:
Tests configs for actions

SmolinIvan committed:
Jest tests finished

SmolinIvan commented on package.json:
```diff
@@ -72,6 +74,7 @@
     "jest": "^29.7.0",
     "jest-css-modules-transform": "^4.4.2",
     "jest-environment-jsdom": "^29.7.0",
+    "jest-localstorage-mock": "^2.4.26",
```
Пришлось добавить, чтобы jest не ругался на localstorage.setItem

SmolinIvan commented on .github/workflows/tests.yml:
Настроил запуск тестов в GitHub action. На запуск стенда под ПР что-то уже не было сил(

SmolinIvan commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,249 @@
+    bottomBun.contains(`${mockIngredients[0].name} (низ)`);
+  });
+
+  // такой поиск безопаснее и нет необходимости добавлять дополнительные теги в компонент
```
Тут объяснил, что вроде как добавление лишних атрибутов в компонент "плохая практика". Учитывая что вебпак всегда генерит разные классы для компонентов, чтобы было к ним еще труднее обратиться, то так безопасность только повышается

SmolinIvan commented on cypress.config.ts:
```diff
@@ -0,0 +1,14 @@
+  e2e: {
+    baseUrl: 'http://localhost:4000',
+    env: {
+      apiUrl: process.env.BURGER_API_URL
```
Оказалось, чтобы Cypress дергал верную ручку (а не localhost:4000/api/*) в CI, надо еще донастраивать

SmolinIvan commented on jest.config.ts:
```diff
@@ -0,0 +1,226 @@
+  // ],
+
+  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
+  moduleNameMapper: {
```
Для jest еще необходимо настраивать импорты

kyzinatra commented on src/services/store.test.ts:
```diff
@@ -0,0 +1,1237 @@
+        user,
+        errorText
+      } = newStore.getState().user;
+
```
Можно лучше: стоит поразбивать тесты по разным файлам, так их будет проще читать и контролировать

kyzinatra commented on src/services/store.test.ts:
```diff
@@ -0,0 +1,1237 @@
+      expect(isAwaiting).toBeFalsy();
+      expect(isAuthed).toBeTruthy();
+      expect(accessToken).toEqual(authorizedUserState.accessToken);
+      expect(refreshToken).toEqual(authorizedUserState.refreshToken);
```
Кажется, чуть проще будет не каждое поле сравнивать а весь объект сразу `toEqual` другому объекту

kyzinatra commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,249 @@
+      mockIngredients[randomNumber].image_large
+    );
+    cy.get('#modals').contains(mockIngredients[randomNumber].name);
+    cy.get('#modals').contains(mockIngredients[randomNumber].calories);
```
Можно лучше: обычно принято в компонентах указывать `data-testid`. Так проще обращаться к конкретным сущностям

kyzinatra commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,249 @@
+  it('Закрытие модального окна на оверлей', function () {
+    cy.get('#modals div').should('not.exist');
+    const randomNumber = Math.floor(
+      Math.random() * (mockIngredients.length - 1)
```
Тесты обычно принято писать детерменированными. То есть без случайностей. Так потенциальный баг проще будет воспроизвести

kyzinatra commented:
Очень хорошая работа. Тесты написаны аккуратно без лишнего кода. Все что нужно вынесено в отдельную константу. Код приятно читать. Все тесты проходят. Хоть тесты и не самая сложная тема, но все равно, вы отлично с ней справились. И продемонстрировали свое знание тестирования и умение покрыть приложение тестами, но все же, рекомендую прочитать мои комментарии к работе, чтобы улучшить ее.
Ваша работа принята.
Успехов в дальнейшем обучении!
